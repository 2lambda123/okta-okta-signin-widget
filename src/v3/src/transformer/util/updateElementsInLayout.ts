/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type MapFn = (uischema: UISchemaElement) => UISchemaElement;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: UISchemaLayout;
  predicate: PredicateFn;
  callback: CallbackFn;
};

type OptionsOld = {
  layout: UISchemaLayout;
  mapFn: MapFn;
  predicateFn: PredicateFn;
};

export const updateElementsInLayout = (options: OptionsOld) => {
  const fn = (
    layout: UISchemaLayout,
    mapFn: MapFn,
    predicateFn: PredicateFn = () => true,
  ): UISchemaLayout => {
    // eslint-disable-next-line no-param-reassign
    layout.elements = layout.elements.map((element) => {
      const { type } = element;
      if (type === UISchemaLayoutType.STEPPER) {
        // eslint-disable-next-line no-param-reassign
        (element as StepperLayout).elements = (element as StepperLayout).elements
          .map((el) => fn(el, mapFn, predicateFn));
        return element;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        return fn(element as UISchemaLayout, mapFn, predicateFn);
      }

      if (predicateFn(element)) {
        return mapFn(element);
      }

      return element;
    });

    return layout;
  };

  return fn(options.layout, options.mapFn, options.predicateFn);
};

export const traveseLayout = (options: Options) => {
  const fn = (
    layout: UISchemaLayout,
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
        return;
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
}
