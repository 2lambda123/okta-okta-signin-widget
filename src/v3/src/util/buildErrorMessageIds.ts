/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const buildErrorMessageIds = (errors: string[], fieldName: string): string => {
  if (errors.length === 1) {
    return `${fieldName}-error`;
  }
  const ids = errors.map((_: string, index: number) => `${fieldName}-error-${index}`);
  return ids.join(' ');
};
