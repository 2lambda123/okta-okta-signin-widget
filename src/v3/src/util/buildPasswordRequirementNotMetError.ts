/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetError = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
): string | undefined => {
  // TODO: Add ticket here to create new translation key
  const errorMessagePrefix = loc('registration.error.password.passwordRequirementsNotMet.prefix', 'login');
  const errorMessageSuffix = passwordRequirements.map((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    return ruleValue === false ? requirement.label : undefined;
  }).filter(Boolean).join(', ') || undefined;

  return typeof errorMessageSuffix !== 'undefined'
    ? `${errorMessagePrefix} ${errorMessageSuffix}`
    : undefined;
};
