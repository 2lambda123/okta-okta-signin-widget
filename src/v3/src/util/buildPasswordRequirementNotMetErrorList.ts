/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessageWithName, ListItem, PasswordValidation } from '../types';
import { loc } from './locUtil';

export const buildPasswordRequirementNotMetErrorList = (
  passwordRequirements: ListItem[],
  passwordValidations: PasswordValidation,
  fieldName: string,
): IdxMessageWithName[] => {
  const errorMessages: IdxMessageWithName[] = [];

  passwordRequirements.forEach((requirement: ListItem) => {
    const ruleValue = passwordValidations[requirement.ruleKey];
    if (ruleValue === false) {
      errorMessages.push({
        name: fieldName,
        class: 'ERROR',
        message: requirement.label,
        i18n: { key: '' },
      });
    }
  });

  if (errorMessages.length > 0) {
    // TODO: Add ticket here to create new translation key that includes colon
    errorMessages.unshift({
      name: fieldName,
      class: 'ERROR',
      message: loc('registration.error.password.passwordRequirementsNotMet', 'login'),
      i18n: { key: 'registration.error.password.passwordRequirementsNotMet' },
    });
  }

  return errorMessages;
};
