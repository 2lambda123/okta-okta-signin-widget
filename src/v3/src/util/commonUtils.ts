/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

// eslint-disable-next-line
export const getDisplayName = (WrappedComponent: any): string => (
  WrappedComponent.displayName || WrappedComponent.name || 'Component'
);

export const handleFormFieldChange = (
  setDirtyFn: ((isDirty: boolean) => void) | undefined,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleChange: (path: string, value: any) => void,
  path: string,
  // eslint-disable-next-line
  value: any,
): void => {
  setDirtyFn?.(true);
  handleChange(path, value);
};
