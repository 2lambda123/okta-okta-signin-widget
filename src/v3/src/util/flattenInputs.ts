/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';

export const flattenInputs = (input: Input, name = ''): Input[] => {
  const res: Input[] = [];
  const { value } = input;

  if (Array.isArray(value)) {
    return value.reduce((acc, curr) => [
      ...acc,
      ...flattenInputs(curr, input.name),
    ], res);
  }

  return [{
    ...input,
    name: name ? `${name}.${input.name}` : input.name,
  }];
};
