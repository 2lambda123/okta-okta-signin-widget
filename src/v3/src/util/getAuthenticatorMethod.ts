/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

export const getAuthenticatorMethod = (
  transaction: IdxTransaction,
): string | undefined => {
  // currentAuthenticator is from enrollment flows and currentAuthenticatorEnrollment is from verify flows
  // @ts-ignore Add currentAuthenticatorEnrollment to IdxContext type
  const { context: { currentAuthenticator, currentAuthenticatorEnrollment } } = transaction;
  const enrollMethods = currentAuthenticator?.value?.methods;
  const verifyMethods = currentAuthenticatorEnrollment?.value?.methods;

  return (enrollMethods && enrollMethods[0].type) || (verifyMethods && verifyMethods[0].type);
};
