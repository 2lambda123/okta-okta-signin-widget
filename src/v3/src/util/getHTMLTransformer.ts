/*
 * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { convertNodeToElement, Options } from 'react-html-parser';

export const getHTMLTransformer = (
  currentNodeName: string,
  newNodeName: string,
  attributes?: Record<string, unknown>,
): Options['transform'] => {
  const transform: Options['transform'] = (node, index) => {
    if (node.type === 'tag' && node.name === currentNodeName) {
      // eslint-disable-next-line no-param-reassign
      node.name = newNodeName;
      if (typeof attributes !== 'undefined') {
        // eslint-disable-next-line no-param-reassign
        node.attribs = attributes;
      }
      return convertNodeToElement(node, index, transform!);
    }
    return undefined;
  };

  return transform;
};
