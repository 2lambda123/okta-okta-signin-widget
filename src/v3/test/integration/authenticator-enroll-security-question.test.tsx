/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/credential/enroll/securityquestion-enroll-mfa.json';

describe('authenticator-enroll-security-question', () => {
  it('should render form', async () => {
    const { container, findByText } = await setup({ mockResponse });

    await findByText(/Set up security question/);
    await findByText(/What is the food you least liked/);
    expect(container).toMatchSnapshot();
  });

  it('should send correct payload with default selections', async () => {
    const {
      authClient, user, findByTestId, findByText,
    } = await setup({ mockResponse });

    await findByText(/Set up security question/);

    const submitButton = await findByText('Verify', { selector: 'button' });
    const answerEle = await findByTestId('credentials.answer') as HTMLInputElement;

    const answer = 'pizza';
    await user.type(answerEle, answer);

    expect(answerEle.value).toEqual(answer);

    await user.click(submitButton);
    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      'POST',
      'https://oie-4695462.oktapreview.com/idp/idx/challenge/answer',
      {
        data: JSON.stringify({
          credentials: {
            questionKey: 'disliked_food',
            answer,
          },
          stateHandle: 'fake-stateHandle',
        }),
        headers: {
          Accept: 'application/json; okta-version=1.0.0',
          'Content-Type': 'application/json',
          'X-Okta-User-Agent-Extended': 'okta-auth-js/9.9.9',
        },
        withCredentials: true,
      },
    );
  });

  it('should send correct payload with custom selections', async () => {
    const {
      authClient, user, findByTestId, findByText, findByLabelText,
    } = await setup({ mockResponse });

    await findByText(/Set up security question/);

    user.click(await findByLabelText(/Create my own security question/));

    const customQuestionEle = await findByTestId('credentials.question') as HTMLInputElement;
    const answerEle = await findByTestId('credentials.answer') as HTMLInputElement;
    const submitButton = await findByText('Verify', { selector: 'button' });

    const question = 'What is the meaning of life?';
    const answer = '42';
    await user.type(customQuestionEle, question);
    await user.type(answerEle, answer);

    expect(customQuestionEle.value).toEqual(question);
    expect(answerEle.value).toEqual(answer);

    await user.click(submitButton);
    expect(authClient.options.httpRequestClient).toHaveBeenCalledWith(
      'POST',
      'https://oie-4695462.oktapreview.com/idp/idx/challenge/answer',
      {
        data: JSON.stringify({
          credentials: {
            questionKey: 'custom',
            question,
            answer,
          },
          stateHandle: 'fake-stateHandle',
        }),
        headers: {
          Accept: 'application/json; okta-version=1.0.0',
          'Content-Type': 'application/json',
          'X-Okta-User-Agent-Extended': 'okta-auth-js/9.9.9',
        },
        withCredentials: true,
      },
    );
  });
});
