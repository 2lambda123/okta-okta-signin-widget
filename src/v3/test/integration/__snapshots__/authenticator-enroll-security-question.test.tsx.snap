// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticator-enroll-security-question should render form 1`] = `
<div>
  <span>
    <main
      class="ods-4o5zYQ ods-76eppy ods-66Tc65 ods-6lar65 ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs auth-container main-container mainViewContainer"
    >
      <div
        class="ods-4o5zYQ ods-2Se4oq ods-2VFzyK ods-1qUaxm ods-1oujX1 ods-78Q1Pv ods-3x5mmc ods-7mcS6o ods-7lpPkQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs siwContainer"
      >
        <div
          class="okta-sign-in-header auth-header siwHeader authCoinSpacing"
        >
          <h1
            class="ods-4o5zYQ ods-2NLKpK ods-5ZKJUt"
          />
          <div
            class="ods-4o5zYQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs iconContainer authCoinOverlay"
          >
            <mocksvgicon />
          </div>
        </div>
        <div
          class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-7kea9D ods-2famef ods-1qUaxm ods-1oujX1 ods-2i2wfv ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
        >
          <div
            class="ods-4o5zYQ ods-76eppy ods-66Tc65 ods-6lar65 ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs identifier-container"
          >
            <div
              class="ods-4o5zYQ ods-aC3Yei ods-2VFzyK ods-aqyCl6 ods-3kEpst ods-IB8Drx ods-ET3J6o ods-3PJtSR ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs identifierContainer"
            >
              <span
                class="ods-4o5zYQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs userIconContainer"
              >
                <svg
                  class="ods-2icygl"
                  fill="none"
                  role="presentation"
                  viewBox="0 0 16 16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M10 3V5C10 6.10457 9.10457 7 8 7C6.89543 7 6 6.10457 6 5V3C6 1.89543 6.89543 1 8 1C9.10457 1 10 1.89543 10 3ZM5 3C5 1.34315 6.34315 0 8 0C9.65685 0 11 1.34315 11 3V5C11 6.65685 9.65685 8 8 8C6.34315 8 5 6.65685 5 5V3ZM4.1305 10.0742C4.26803 10.0181 4.459 10 5.41376 10H10.5862C11.541 10 11.732 10.0181 11.8695 10.0742C12.0299 10.1398 12.1706 10.2459 12.2777 10.3821C12.3695 10.4989 12.4393 10.6776 12.7016 11.5956L13.6743 15H2.32573L3.29841 11.5956C3.5607 10.6776 3.63054 10.4989 3.72233 10.3821C3.82941 10.2459 3.97006 10.1398 4.1305 10.0742ZM13.6631 11.3209L14.7143 15L15 16H13.96H2.04002H1L1.28571 15L2.33689 11.3209C2.57458 10.489 2.69343 10.073 2.93606 9.76423C3.15022 9.49171 3.43152 9.27953 3.75239 9.14848C4.11592 9 4.54854 9 5.41376 9H10.5862C11.4515 9 11.8841 9 12.2476 9.14848C12.5685 9.27953 12.8498 9.49171 13.0639 9.76423C13.3066 10.073 13.4254 10.489 13.6631 11.3209Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <span
                class="ods-4o5zYQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs identifier identifierSpan"
                data-se="identifier"
              >
                tester14@test.com
              </span>
            </div>
          </div>
          <form
            novalidate=""
          >
            <div
              class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
            >
              <div
                class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
              >
                <div
                  class="ods-4o5zYQ ods-76eppy ods-2famef ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                >
                  <h2
                    class="ods-4o5zYQ ods-2NLKpK ods-5ZKJUt ods-4vLZvh"
                  >
                    Set up security question
                  </h2>
                </div>
              </div>
              <div
                class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
              >
                <fieldset
                  class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-4zKnH2 ods-6uQyQM"
                >
                  <legend
                    class="ods-4o5zYQ ods-76eppy ods-5VQ9Rl ods-7KS5Kt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-g8inGu"
                    htmlfor="questionType"
                  >
                    <span
                      class="ods-4o5zYQ ods-nqMobL ods-5fvoER ods-6rDd3P ods-7KlglL ods-6MIfC2 ods-14pjkk ods-5IIKpe ods-6iOjmZ"
                    >
                      
                    </span>
                  </legend>
                  <div
                    class="ods-4o5zYQ ods-15myrS ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-7tvTLl"
                  >
                    <input
                      aria-describedby="questionType-error"
                      class="ods-4B2VQs"
                      id="generated"
                      name="questionType"
                      type="radio"
                      value="predefined"
                    />
                    <label
                      class="ods-5OMySi"
                      for="generated"
                    >
                      Choose a security question
                    </label>
                  </div>
                  <div
                    class="ods-4o5zYQ ods-15myrS ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-7tvTLl"
                  >
                    <input
                      aria-describedby="questionType-error"
                      class="ods-4B2VQs"
                      id="generated"
                      name="questionType"
                      type="radio"
                      value="custom"
                    />
                    <label
                      class="ods-5OMySi"
                      for="generated"
                    >
                      Create my own security question
                    </label>
                  </div>
                </fieldset>
              </div>
              <div
                class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
              >
                <div
                  class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                >
                  <div
                    class="ods-4o5zYQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                  >
                    <div
                      class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-4zKnH2"
                    >
                      <label
                        class="ods-4o5zYQ ods-76eppy ods-5VQ9Rl ods-7KS5Kt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-g8inGu"
                        for="generated"
                      >
                        <span
                          class="ods-4o5zYQ ods-nqMobL ods-5fvoER ods-6rDd3P ods-7KlglL ods-6MIfC2 ods-14pjkk ods-5IIKpe ods-6iOjmZ"
                        >
                          Choose a security question
                        </span>
                      </label>
                      <div
                        class="ods-mUkyrD"
                      >
                        <select
                          class="ods-1B2Zfy"
                          id="generated"
                          name="credentials.questionKey"
                        >
                          <option
                            value=""
                          />
                          <option
                            value="disliked_food"
                          >
                            What is the food you least liked as a child?
                          </option>
                          <option
                            value="name_of_first_plush_toy"
                          >
                            What is the name of your first stuffed animal?
                          </option>
                          <option
                            value="first_award"
                          >
                            What did you earn your first medal or award for?
                          </option>
                          <option
                            value="favorite_security_question"
                          >
                            What is your favorite security question?
                          </option>
                          <option
                            value="favorite_toy"
                          >
                            What is the toy/stuffed animal you liked the most as a kid?
                          </option>
                          <option
                            value="first_computer_game"
                          >
                            What was the first computer game you played?
                          </option>
                          <option
                            value="favorite_movie_quote"
                          >
                            What is your favorite movie quote?
                          </option>
                          <option
                            value="first_sports_team_mascot"
                          >
                            What was the mascot of the first sports team you played on?
                          </option>
                          <option
                            value="first_music_purchase"
                          >
                            What music album or song did you first purchase?
                          </option>
                          <option
                            value="favorite_art_piece"
                          >
                            What is your favorite piece of art?
                          </option>
                          <option
                            value="grandmother_favorite_desert"
                          >
                            What was your grandmother's favorite dessert?
                          </option>
                          <option
                            value="first_thing_cooked"
                          >
                            What was the first thing you learned to cook?
                          </option>
                          <option
                            value="childhood_dream_job"
                          >
                            What was your dream job as a child?
                          </option>
                          <option
                            value="place_where_significant_other_was_met"
                          >
                            Where did you meet your spouse/significant other?
                          </option>
                          <option
                            value="favorite_vacation_location"
                          >
                            Where did you go for your favorite vacation?
                          </option>
                          <option
                            value="new_years_two_thousand"
                          >
                            Where were you on New Year's Eve in the year 2000?
                          </option>
                          <option
                            value="favorite_speaker_actor"
                          >
                            Who is your favorite speaker/orator?
                          </option>
                          <option
                            value="favorite_book_movie_character"
                          >
                            Who is your favorite book/movie character?
                          </option>
                          <option
                            value="favorite_sports_player"
                          >
                            Who is your favorite sports player?
                          </option>
                        </select>
                        <span
                          class="ods-7mdg51"
                        >
                          <svg
                            class="ods-2icygl"
                            fill="none"
                            role="presentation"
                            viewBox="0 0 16 16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M8 10.2929L12.6464 5.64645L13.3536 6.35355L8.35355 11.3536C8.15829 11.5488 7.84171 11.5488 7.64645 11.3536L2.64645 6.35355L3.35355 5.64645L8 10.2929Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                >
                  <div
                    class="ods-4o5zYQ ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                  >
                    <div
                      class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                    >
                      <div
                        class="ods-4o5zYQ ods-76eppy ods-2Se4oq ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-4zKnH2"
                      >
                        <label
                          class="ods-4o5zYQ ods-76eppy ods-5VQ9Rl ods-7KS5Kt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs ods-g8inGu"
                          for="generated"
                        >
                          <span
                            class="ods-4o5zYQ ods-nqMobL ods-5fvoER ods-6rDd3P ods-7KlglL ods-6MIfC2 ods-14pjkk ods-5IIKpe ods-6iOjmZ"
                          >
                            Answer
                          </span>
                        </label>
                        <div
                          class="ods-10aW0m"
                        >
                          <input
                            class="ods-7qBw9I"
                            data-se="credentials.answer"
                            id="generated"
                            name="credentials.answer"
                            type="password"
                          />
                        </div>
                        
                      </div>
                    </div>
                    <button
                      class="ods-4o5zYQ ods-2NLKpK ods-5E7TSu ods-4vGTAP"
                      data-se="show-password"
                      type="button"
                    >
                      <span
                        class="ods-IkfPY2"
                      >
                        Show Password
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
                >
                  <button
                    class="ods-4o5zYQ ods-2NLKpK ods-5E7TSu ods-6Tybtd"
                    data-se="#/properties/submit"
                    data-type="save"
                    type="submit"
                  >
                    <span
                      class="ods-IkfPY2"
                    >
                      Verify
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
              >
                <button
                  class="ods-4o5zYQ ods-2NLKpK ods-5E7TSu ods-3DL1tX"
                  data-se="switchAuthenticator"
                  type="button"
                >
                  <span
                    class="ods-IkfPY2"
                  >
                    Return to authenticator list
                  </span>
                </button>
              </div>
              <div
                class="ods-4o5zYQ ods-1OsrNt ods-2oc5SK ods-4XAbls ods-2NLKpK ods-1ll9ky ods-4IdjLs"
              >
                <button
                  class="ods-4o5zYQ ods-2NLKpK ods-5E7TSu ods-3DL1tX"
                  data-se="cancel"
                  type="button"
                >
                  <span
                    class="ods-IkfPY2"
                  >
                    Back to sign in
                  </span>
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </main>
  </span>
</div>
`;
